File: app/src/main/java/com/example/myapplication/main/data/remote/retrofit/RetrofitClientImpl.java
package com.example.myapplication.main.data.remote.retrofit;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.example.myapplication.main.Constants.BASE_URL;

public class RetrofitClientImpl implements RetrofitClient {

    @Override
    public Retrofit createRetrofit() {
        return new retrofit2.Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
    }
}

File: app/src/main/java/com/example/myapplication/main/data/remote/retrofit/RetrofitClientImpl.kt
package com.example.myapplication.main.data.remote.retrofit

import com.example.myapplication.main.Constants
import com.example.myapplication.main.Constants.BASE_URL
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class RetrofitClientImpl : RetrofitClient {

    override fun createRetrofit(): Retrofit {
        return retrofit2.Retrofit.Builder()
                .client(okHttpClient())
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
    }

    private fun okHttpClient(): OkHttpClient {
        return OkHttpClient.Builder()
                .addInterceptor(authInterceptor)
                .build()
    }

    //Creating Auth Interceptor to add api_key query in front of all the requests.
    private val authInterceptor = Interceptor { chain ->
        val newUrl = chain.request().url()
                .newBuilder()
                .addQueryParameter("api_key", Constants.API_KEY)
                .build()

        val newRequest = chain.request()
                .newBuilder()
                .url(newUrl)
                .build()

        chain.proceed(newRequest)
    }
}